marlowe-playground-client/src/Wallet.purs

```diff
@@ -0,0 +1,1301 @@
+
+handleAction (ChangeCurrencyInput token money) = modifying _assetChanges (Map.insert token money)
+
+handleAction (AddCurrency token) = do
+  newMoney <- use (_assetChanges <<< to (Map.lookup token) <<< to (fromMaybe zero))
+  modifying (_openWallet <<< _Just <<< _assets) (Map.alter (addMoney newMoney) token)
+  where
+  addMoney newMoney (Just oldMoney) = Just $ oldMoney + newMoney
+
+  addMoney newMoney _ = Just newMoney
+
+handleAction (RenameWallet name) = assign (_openWallet <<< _Just <<< _name) name
+
+handleAction CreateWallet = do
+  idx <- use (_wallets <<< to Map.size <<< to (add one))
+  let
+    name = "Wallet " <> show idx
+
+    wallet = Wallet { name, assets: mempty, loadedContract: Nothing }
+  modifying _wallets (Map.insert name wallet)
+  assign _openWallet (Just wallet)
+  assign _view $ WalletView $ wallet ^. _name
+
+handleAction (SelectContract contract) = do
+  assign _walletLoadedContract (Just contract)
+  assign _addInputError Nothing
+
+handleAction (ChangeHelpContext help) = do
+  assign _helpContext help
+  scrollHelpPanel
+
+scrollHelpPanel :: forall m. MonadEffect m => HalogenM State Action ChildSlots Message m Unit
+scrollHelpPanel =
+  liftEffect do
+    window <- Web.window
+    document <- toDocument <$> W.document window
+    mSidePanel <- WC.item 0 =<< D.getElementsByClassName "sidebar-composer" document
+    case mSidePanel of
+      Nothing -> pure unit
+      Just sidePanel -> do
+        scrollHeight <- E.scrollHeight sidePanel
+        setScrollTop scrollHeight sidePanel
+
+render :: forall p. State -> HTML p Action
+render state =
+  div []
+    [ section [ classes [ panelSubHeader, aHorizontal ] ]
+        [ div [ classes [ panelSubHeaderMain, aHorizontal ] ]
+            [ div [ classes [ ClassName "wallet-title", aHorizontal, jFlexStart ] ]
+                [ div [ classes [ ClassName "demos", spaceLeft ] ]
+                    [ small_ [ text "Created Wallets:" ]
+                    ]
+                ]
+            , ul [ classes [ ClassName "wallet-list", aHorizontal ] ]
+                (map (displayWallet walletName) wallets)
+            , addWalletButton
+            , a [ class_ (ClassName "clear-all-contracts"), onClick $ Just <<< const ResetAll ] [ text "Clear All" ]
+            ]
+        , div [ classes [ panelSubHeaderSide, expanded (state ^. _showRightPanel) ] ]
+            [ a [ classes [ (ClassName "drawer-icon-click") ], onClick $ const $ Just $ ShowRightPanel (state ^. (_showRightPanel <<< to not)) ]
+                [ img [ src closeDrawerIcon, class_ (ClassName "drawer-icon") ] ]
+            ]
+        ]
+    , section [ class_ (ClassName "code-panel") ]
+        [ mainPanel state
+        , rightPanel state
+        ]
+    ]
+  where
+  isRunning = fromMaybe false $ state ^? (_walletLoadedContract <<< _Just <<< _started)
+
+  walletName = fromMaybe "No Wallet Open" $ state ^? (_openWallet <<< _Just <<< _name)
+
+  wallets = map snd $ Map.toUnfoldable $ state ^. _wallets
```

Comment:
What about `wallets = toArrayOf (_wallets <<< traversed) state`

Reply:
it's even in `Data.Lens.Fold` in the version of profunctor-lenses I'm now using!